### DISCUSSION TITLE
improve send selection

### DISCUSSION THREAD

**Discussion by yuuuxt** [ionide/ionide-vscode-fsharp#1728] [General]
I created a simple PR #1723 to improve the behaviour of `fsi:send selection`.

I also checked other F# IDEs about this:
- Visual Studio: No such feature. Can be achieved with `TextMacros` extension (comment&uncomment to select the complete lines, and need to add custom command to trigger send selection);
- Rider: No such feature. Can be achieved by adding a macro (`EditorSelectLine` & `SendToFsi`; easier than VS)

For me, I don't see any drawback of making such a change, and please correct me if that's wrong. 

What do you think? Thanks

**Comment by jasiozet** [comment]
I think that the idea of improving selection is a great one, but worth noting that with this Ionide looses a feature of selecting something mid line and evaluating it. I'm not sure if it is a big deal.

functionWithTwoArguments (anotherFunction arg) (anotherFunction2 arg2)
I could imagine that I want to select one of the parens and evaluate the middle, with that change it would always evaluate the whole line.

My idea of solving this problem, that I want to explore is to provide a shortcut for evaluating paragraphs.

let s = 0

let a = 1 // middle
let b = 1 // section
printfn "%d" <| a + b // middle section

let c = 1
Placing a cursor anywhere in the middle section (commented) and evaluating would send the:

let a = 1
let b = 1
printfn "%d" <| a + b
To REPL

This comment was copied from https://github.com/ionide/ionide-vscode-fsharp/pull/1723

I think the current best idea I have is to have different options:
* current behavior (line selection)
* yuuuxt's suggestion here
* paragraph suggestion by me
* possibly others

And pick in the config.

For me I like one shortcut sending to repl and I want to send the whole multiline functions (that's why I would love paragraph one). One line is also sometimes useful for me. 

I am also incredibly curious what other people experience and how the leverage the REPL for their benefits

**Comment by Krzysztof-Cieslak** [comment]
I don't think we should change "send selection" command - ability to send only the part of the line seems to be important due to referential transparency.

However, how about updating the "send the line" command to the behaviour proposed in PR? Send line currently checks only the start point of the selection and sends the single line, but expanding it to all the lines that are (even partially) covered by selection seems reasonable to me.

@baronfel any opinion?

**Reply by baronfel** [reply]
I like that a lot more. I was sceptical of changing 'send selection' for that same reason.

**Reply by jasiozet** [reply]
How about having different commands for these things so people can bind them however they want?
Send line
Send selection
Send paragraph
Send smart selection (mentioned PR)

**Reply by yuuuxt** [reply]
Maybe for one shortcut we can have:
- no selection: send the single line
- selection in single line: send selection only
- selection in multiple lines: send the complete multiple lines

**Reply by jasiozet** [reply]
I like this idea + new command for send paragraph 😉

**Comment by yuuuxt** [comment]
update after > 2 years - #2051 is just created to have a send selection extended to whole line command.
